### PR TITLE
[ISSUE #94] Feat sequence

### LINKED ISSUE #94: Simple sequence implementation.
**What would you like to be added**:
Generate pk when executing `INSERT`:
1. register entrance
2. impls: snowflake/sequence table based/...

**Why is this needed**:

### PR DESCRIPTION
**Pull request by chuntaojun**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #94 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
- 支持已snowflake为底座的自增ID插件
```

### CODE REVIEW

File: pkg/proto/sequence.go [hunk 1]
```diff
@@ -22,14 +22,62 @@ import (
...
+var (
+	createSequenceManager func() SequenceManager
+)
+
+func SetSequenceManagerCreator(creator func() SequenceManager) {
+	createSequenceManager = creator
+}
+
+// SequenceSupplier 创建 Sequence 的创建器
+type SequenceSupplier func() EnchanceSequence
```
Comment (dongzl) [inline pkg/proto/sequence.go:38]: Suggest document in English.

File: pkg/runtime/namespace/namespace.go [hunk 2]
```diff
@@ -93,10 +95,11 @@ type (
...
 func New(name string, optimizer proto.Optimizer, commands ...Command) *Namespace {
 	ns := &Namespace{
-		name:      name,
-		optimizer: optimizer,
-		cmds:      make(chan Command, 1),
-		done:      make(chan struct{}),
+		name:        name,
+		optimizer:   optimizer,
+		sequenceMgn: proto.GetSequenceManager(),
+		cmds:        make(chan Command, 1),
```
Comment (dongzl) [inline pkg/runtime/namespace/namespace.go:101]: Suggest not use shorthand `sequenceMgn`

File: scripts/sharding.sql [hunk 3]
```diff
@@ -500,4 +500,4 @@ CREATE TABLE IF NOT EXISTS `employees_0003`.`student_0031`
     UNIQUE KEY `uk_uid` (`uid`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
 
-INSERT INTO employees_0000.student_0001 VALUES (1, 1, 'scott', 95, 'nc_scott', 0, 16, NOW(), NOW());
+INSERT INTO employees_0000.student_0001 VALUES (1, 1, 'scott', 95, 'nc_scott', 0, 16, NOW(), NOW());
```
Comment (dongzl) [inline scripts/sharding.sql:503]: Suggest add blank line at end of file.

File: scripts/sequence.sql [hunk 4]
```diff
@@ -28,4 +28,4 @@ CREATE TABLE IF NOT EXISTS `employees_0000`.`sequence`
     PRIMARY KEY (`id`),
     UNIQUE KEY `uk_name` (`name`)
 ) ENGINE = InnoDB
-  DEFAULT CHARSET = utf8mb4;
+  DEFAULT CHARSET = utf8mb4;
```
Comment (dongzl) [inline scripts/sequence.sql:31]: Suggest add blank line at end of file.

File: scripts/init.sql [hunk 5]
```diff
@@ -110,4 +110,4 @@ CREATE TABLE salaries (
     FOREIGN KEY (emp_no) REFERENCES employees (emp_no) ON DELETE CASCADE,
     PRIMARY KEY (emp_no, from_date)
 )
-;
+;
```
Comment (dongzl) [inline scripts/init.sql:113]: Suggest add blank line at end of file.

File: pkg/sequence/snowflake/snowflake.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+const (
+	SequencePluginName = "snowflake"
+)
+
+const (
+	_initTableSql = `
+	CREATE TABLE IF NOT EXISTS __arana_snowflake_sequence (
+		work_id int AUTO_INCREMENT COMMENT '主键',
```
Comment (dongzl) [inline pkg/sequence/snowflake/snowflake.go:47]: Suggest doc in English.

File: pkg/sequence/sequence.go [hunk 7]
```diff
@@ -17,4 +17,92 @@
...
+
+// SequenceManager 统一管理 Seqneuce 的 Manager
+type sequenceManager struct {
+	lock             sync.RWMutex
+	sequenceOptions  map[string]proto.SequenceConfig
+	sequenceRegistry map[string]proto.EnchanceSequence
+}
+
+// CreateSequence
+func (sMgn *sequenceManager) CreateSequence(ctx context.Context, conn proto.VConn, conf proto.SequenceConfig) (proto.Sequence, error) {
```
Comment (dongzl) [inline pkg/sequence/sequence.go:58]: Suggest not use shorthand `sMgn`,  `manager` is OK?

File: pkg/sequence/group/group.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+	SequencePluginName = "group"
+)
+
+type groupSequence struct {
+	workdId      int32
+	curentVal    int64
+	preTimestamp int64
+}
+
+// Start 启动 Sequence，做一些初始化操作
```
Comment (dongzl) [inline pkg/sequence/group/group.go:45]: Suggest in English.

File: pkg/proto/rule/rule.go [hunk 9]
```diff
@@ -55,8 +55,22 @@ const (
...
+	autoIncrement AutoIncrement
+	topology      *Topology
+	shards        map[string][2]*ShardMetadata // column -> [db shard metadata,table shard metadata]
+}
+
+func (vt *VTable) SetDefaultAutoIncrement() {
+	vt.autoIncrement = AutoIncrement{
+		Type: "snowflake",
+		Option: map[string]string{
+			"": "",
```
Comment (dongzl) [inline pkg/proto/rule/rule.go:67]: Why init `"": ""`,  `make(map[string]string)`?

File: pkg/proto/sequence.go [hunk 10]
```diff
@@ -22,14 +22,62 @@ import (
...
+
 // Sequence represents a global unique id generator.
-type Sequence interface {
-	// Next generates a next value in int64.
-	Next(ctx context.Context) (int64, error)
+type EnchanceSequence interface {
+	Sequence
+	// Start
+	Start(ctx context.Context, option SequenceConfig) error
+	//
```
Comment (dongzl) [inline pkg/proto/sequence.go:65]: Suggest remove this no meaning doc.

File: pkg/runtime/namespace/namespace.go [hunk 11]
```diff
@@ -76,6 +76,8 @@ type (
 
 		name string // the name of Namespace
 
+		sequenceMgn proto.SequenceManager
```
Comment (dongzl) [inline pkg/runtime/namespace/namespace.go:79]: Suggest not use shorthand `sequenceMgn`

**Review by dongzl — CHANGES_REQUESTED** [review]

File: pkg/sequence/group/group.go [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+	})
+}
+
+const (
+	SequencePluginName = "group"
+)
+
+type groupSequence struct {
+	workdId      int32
+	curentVal    int64
```
Comment (AlexStocks) [inline pkg/sequence/group/group.go:41]: currentVal

File: pkg/runtime/namespace/namespace.go (hunk 11 above)
Comment (AlexStocks) [inline pkg/runtime/namespace/namespace.go:79]: sequenceMgr

File: conf/config.yaml [hunk 13]
```diff
@@ -82,6 +82,9 @@ data:
     tables:
       - name: employees.student
         allow_full_scan: true
+        auto_increment:
```
Comment (jjeffcaii) [inline conf/config.yaml:85]: How about using `sequence` instead of `auto_increment`?

File: pkg/proto/rule/rule.go [hunk 14]
```diff
@@ -55,8 +53,20 @@ const (
 // VTable represents a virtual/logical table.
 type VTable struct {
 	attributes
-	topology *Topology
-	shards   map[string][2]*ShardMetadata // column -> [db shard metadata,table shard metadata]
+	autoIncrement AutoIncrement
```
Comment (jjeffcaii) [inline pkg/proto/rule/rule.go:56]: Use pointer, use default if nil, no need to set default value manually.

File: pkg/proto/rule/rule.go [hunk 15]
```diff
@@ -55,8 +53,20 @@ const (
...
+func (vt *VTable) SetDefaultAutoIncrement() {
+	vt.autoIncrement = AutoIncrement{
+		Type:   "snowflake",
+		Option: make(map[string]string),
+	}
+}
+
+func (vt *VTable) GetAutoIncrement() AutoIncrement {
+	return vt.autoIncrement
 }
```
Comment (jjeffcaii) [inline pkg/proto/rule/rule.go:70]: Use pointer, use default if nil, no need to set default value manually.

File: pkg/proto/sequence.go [hunk 16]
```diff
@@ -22,14 +22,62 @@ import (
 	"context"
 )
 
+var (
+	ContextVconnKey = struct{}{}
```
Comment (jjeffcaii) [inline pkg/proto/sequence.go:26]: Should use type alias, for example:
```go
type VConnCtxKey struct {
}

context.WithValue(ctx, VConnCtxKey{}, ...)



```

File: pkg/runtime/namespace/namespace.go [hunk 17]
```diff
@@ -22,17 +22,12 @@ import (
 	"io"
 	"sort"
 	"sync"
-)
 
-import (
+	"github.com/arana-db/arana/pkg/proto"
+	"github.com/arana-db/arana/pkg/proto/rule"
```
Comment (jjeffcaii) [inline pkg/runtime/namespace/namespace.go:27]: format imports

File: pkg/sequence/snowflake/snowflake.go [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+		PRIMARY KEY (work_id),
+		UNIQUE KEY(node_id, table_name)
+	) ENGINE = InnoDB;
+	`
+
+	_getWorkId = `INSERT IGNORE INTO __arana_snowflake_sequence(node_id, table_name) VALUE (?, ?)`
+)
+
+var (
+	mu sync.Mutex
```
Comment (jjeffcaii) [inline pkg/sequence/snowflake/snowflake.go:59]: Why declare a global mutex here?

**Review by jjeffcaii — APPROVED** [review]

File: conf/config.yaml (hunk 13 above)
Comment (chuntaojun) [reply conf/config.yaml:85]: i think， auto

File: pkg/sequence/snowflake/snowflake.go [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+
+const (
+	_initTableSql = `
+	CREATE TABLE IF NOT EXISTS __arana_snowflake_sequence (
+		work_id int AUTO_INCREMENT COMMENT '主键',
+		node_id varchar(255) NOT NULL COMMENT '节点的唯一标识',
+		table_name varchar(255) NOT NULL COMMENT 'arana中的逻辑表名称',
+		PRIMARY KEY (work_id),
+		UNIQUE KEY(node_id, table_name)
+	) ENGINE = InnoDB;
```
Comment (dongzl) [inline pkg/sequence/snowflake/snowflake.go:51]: Suggest doc in English.

**Review by dongzl — CHANGES_REQUESTED** [review]

**Review by dongzl — APPROVED** [review]

File: pkg/runtime/optimize/optimizer.go [hunk 20]
```diff
@@ -808,8 +812,55 @@ func (o optimizer) rewriteSelectStatement(ctx context.Context, conn proto.VConn,
 	return nil
 }
 
+func (o optimizer) createSequenceIfAbsent(ctx context.Context, conn proto.VConn, vTableName string, tableMetadata *proto.TableMetadata) error {
+
+	var (
+		ru = rcontext.Rule(ctx)
```
Comment (baerwang) [inline pkg/runtime/optimize/optimizer.go:818]: 不要放在判断为空前初始化

File: pkg/sequence/sequence.go [hunk 21]
```diff
@@ -17,4 +17,92 @@
...
+type sequenceManager struct {
+	lock             sync.RWMutex
+	sequenceOptions  map[string]proto.SequenceConfig
+	sequenceRegistry map[string]proto.EnchanceSequence
+}
+
+// CreateSequence create one sequence instance
+func (m *sequenceManager) CreateSequence(ctx context.Context, conn proto.VConn, conf proto.SequenceConfig) (proto.Sequence, error) {
+	m.lock.RLock()
+	if seq, exist := m.sequenceRegistry[conf.Name]; exist {
```
Comment (baerwang) [inline pkg/sequence/sequence.go:60]:
```
m.lock.RLock()
m.lock.RUnlock()
```

File: pkg/sequence/sequence.go [hunk 22]
```diff
@@ -17,4 +17,92 @@
...
+		return &sequenceManager{
+			sequenceOptions:  make(map[string]proto.SequenceConfig),
+			sequenceRegistry: make(map[string]proto.EnchanceSequence),
+		}
+	})
+}
+
+// SequenceManager Uniform management of seqneuce manager
+type sequenceManager struct {
+	lock             sync.RWMutex
```
Comment (baerwang) [inline pkg/sequence/sequence.go:52]: 可以不用定义变量
```go
type sequenceManager struct {
    sync.RWMutex
}

Comment (chuntaojun) [reply pkg/sequence/sequence.go:52]: 这个可能不太好？还是倾向于组合的方式来

File: pkg/util/net/net.go [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+package net
+
+import (
+	"net"
+	"os"
+	"strings"
+)
+
+func FindSelfIP() string {
+	openIPV6 := os.Getenv("arana.net.ipv6")
```
Comment (Lvnszn) [inline pkg/util/net/net.go:27]: 这个放下面的 panic 后面可能会更好点

File: pkg/proto/sequence.go [hunk 24]
```diff
@@ -22,14 +22,50 @@ import (
...
+	// Stop stop sequence
+	Stop() error
 }
 
 // Sequencer represents the factory to create a Sequence by table name.
-type Sequencer interface {
-	// Sequence returns the Sequence of table.
-	Sequence(ctx context.Context, table string) (Sequence, error)
+type SequenceManager interface {
+	// CreateSequence create sequence instance
```
Comment (AlexStocks) [inline pkg/proto/sequence.go:61]: create = creates

**Review by AlexStocks — APPROVED** [review]

**Pull request merged**